staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation_test.go

```diff
@@ -0,0 +1,157 @@
+	tests := []struct {
+		name  string
+		input apiextensions.CustomResourceValidation
+	}{
+		{
+			name: "default cel validation",
+			input: apiextensions.CustomResourceValidation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"embedded": {
+							Type: "object",
+							Properties: map[string]apiextensions.JSONSchemaProps{
+								"metadata": {
+									Type:              "object",
+									XEmbeddedResource: true,
+									Properties: map[string]apiextensions.JSONSchemaProps{
+										"name": {
+											Type: "string",
+											XValidations: apiextensions.ValidationRules{
+												{
+													Rule: "self == 'singleton'",
+												},
+											},
+											Default: jsonPtr("singleton"),
+										},
+									},
+								},
+							},
+						},
+						"value": {
+							Type: "string",
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule: "self.startsWith('kube')",
+								},
+							},
+							Default: jsonPtr("kube-everything"),
+						},
+						"object": {
+							Type: "object",
+							Properties: map[string]apiextensions.JSONSchemaProps{
+								"field1": {
+									Type: "integer",
+								},
+								"field2": {
+									Type: "integer",
+								},
+							},
+							XValidations: apiextensions.ValidationRules{
+								{
+									Rule: "self.field1 < self.field2",
+								},
+							},
+							Default: jsonPtr(map[string]interface{}{"field1": 1, "field2": 2}),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			schema := tt.input.OpenAPIV3Schema
+			ss, err := structuralschema.NewStructural(schema)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+
+			f := NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+
+			// cost budget is large enough to pass all validation rules
+			allErrs, err := validate(field.NewPath("test"), ss, ss, f, false, false, math.MaxInt64)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+
+			for _, valErr := range allErrs {
+				t.Errorf("unexpected error: %v", valErr)
+			}
+
+			// cost budget exceeded for all validation rules
+			allErrs, err = validate(field.NewPath("test"), ss, ss, f, false, false, 0)
+			if err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			meetErr := 0
+			for _, valErr := range allErrs {
+				if valErr.Type == field.ErrorTypeInvalid && strings.Contains(valErr.Error(), "validation failed due to running out of cost budget, no further validation rules will be run") {
+					meetErr += 1
+				}
+
+			}
+			if meetErr == 0 {
+				t.Errorf("expect err of running out of cost budget but did not find")
+			}
+			if meetErr != 3 {
+				t.Errorf("expect 3 errs of running out of cost budget returned but get %v errs", meetErr)
```

Comment:
this is surprising... once we exceed our budget, we should stop evaluating further rules, we should get back at most one "budget exceeded" error

Reply:
if this requires threading/propagating `remainingCost` through the default validation function, I'm ok doing this in a follow-up, but it does need to be done